File: src/cudgel/query.py
"""Natural language query functionality."""

from typing import Any, Optional

from cudgel.config import CudgelConfig
from cudgel.database import Database
from cudgel.embeddings import EmbeddingGenerator


class CodeQuery:
    """Query code using natural language."""

    def __init__(self, config: CudgelConfig):
        self.config = config
        self.db = Database(config)
        self.embedder = EmbeddingGenerator(config)

    async def initialize(self) -> None:
        """Initialize the query system."""
        await self.db.connect()
        self.embedder.load_model()

    async def close(self) -> None:
        """Close connections."""
        await self.db.close()

    async def search_symbols(
        self,
        query: str,
        limit: int = 10,
        repository_path: Optional[str] = None
    ) -> list[dict[str, Any]]:
        """
        Search for symbols using natural language query.

        Args:
            query: Natural language query
            limit: Maximum number of results
            repository_path: Optional filter by repository path

        Returns:
            List of matching symbols with metadata
        """
        # Generate query embedding
        query_embedding = self.embedder.encode_query(query)

        if not self.db.conn:
            await self.db.connect()

        # Search for similar symbols using vector similarity
        async with self.db.conn.cursor() as cur:
            if repository_path:
                await cur.execute(
                    """
                    SELECT
                        s.id,
                        s.name,
                        s.kind,
                        s.signature,
                        s.docstring,
                        s.start_line,
                        s.end_line,
                        f.path,
                        f.language,
                        r.path as repo_path,
                        r.name as repo_name,
                        1 - (s.embedding <=> %s::vector) as similarity
                    FROM symbols s
                    JOIN files f ON s.file_id = f.id
                    JOIN repositories r ON f.repository_id = r.id
                    WHERE r.path = %s
                    ORDER BY s.embedding <=> %s::vector
                    LIMIT %s
                    """,
                    (query_embedding.tolist(), repository_path, query_embedding.tolist(), limit),
                )
            else:
                await cur.execute(
                    """
                    SELECT
                        s.id,
                        s.name,
                        s.kind,
                        s.signature,
                        s.docstring,
                        s.start_line,
                        s.end_line,
                        f.path,
                        f.language,
                        r.path as repo_path,
                        r.name as repo_name,
                        1 - (s.embedding <=> %s::vector) as similarity
                    FROM symbols s
                    JOIN files f ON s.file_id = f.id
                    JOIN repositories r ON f.repository_id = r.id
                    ORDER BY s.embedding <=> %s::vector
                    LIMIT %s
                    """,
                    (query_embedding.tolist(), query_embedding.tolist(), limit),
                )

            results = await cur.fetchall()
            return list(results)

    async def search_code(
        self,
        query: str,
        limit: int = 10,
        repository_path: Optional[str] = None
    ) -> list[dict[str, Any]]:
        """
        Search for code chunks using natural language query.

        Args:
            query: Natural language query
            limit: Maximum number of results
            repository_path: Optional filter by repository path

        Returns:
            List of matching code chunks with metadata
        """
        # Generate query embedding
        query_embedding = self.embedder.encode_query(query)

        if not self.db.conn:
            await self.db.connect()

        # Search for similar code chunks
        async with self.db.conn.cursor() as cur:
            if repository_path:
                await cur.execute(
                    """
                    SELECT
                        c.id,
                        c.text,
                        c.start_line,
                        c.end_line,
                        s.name as symbol_name,
                        s.kind as symbol_kind,
                        f.path,
                        f.language,
                        r.path as repo_path,
                        r.name as repo_name,
                        1 - (c.embedding <=> %s::vector) as similarity
                    FROM code_chunks c
                    JOIN files f ON c.file_id = f.id
                    JOIN repositories r ON f.repository_id = r.id
                    LEFT JOIN symbols s ON c.symbol_id = s.id
                    WHERE r.path = %s
                    ORDER BY c.embedding <=> %s::vector
                    LIMIT %s
                    """,
                    (query_embedding.tolist(), repository_path, query_embedding.tolist(), limit),
                )
            else:
                await cur.execute(
                    """
                    SELECT
                        c.id,
                        c.text,
                        c.start_line,
                        c.end_line,
                        s.name as symbol_name,
                        s.kind as symbol_kind,
                        f.path,
                        f.language,
                        r.path as repo_path,
                        r.name as repo_name,
                        1 - (c.embedding <=> %s::vector) as similarity
                    FROM code_chunks c
                    JOIN files f ON c.file_id = f.id
                    JOIN repositories r ON f.repository_id = r.id
                    LEFT JOIN symbols s ON c.symbol_id = s.id
                    ORDER BY c.embedding <=> %s::vector
                    LIMIT %s
                    """,
                    (query_embedding.tolist(), query_embedding.tolist(), limit),
                )

            results = await cur.fetchall()
            return list(results)

    async def search(
        self,
        query: str,
        limit: int = 10,
        repository_path: Optional[str] = None,
        search_type: str = "both"
    ) -> dict[str, Any]:
        """
        Unified search combining symbols and code chunks.

        Args:
            query: Natural language query
            limit: Maximum number of results per type
            repository_path: Optional filter by repository path
            search_type: "symbols", "code", or "both"

        Returns:
            Dictionary with symbols and/or code results
        """
        results: dict[str, Any] = {}

        if search_type in ("symbols", "both"):
            results["symbols"] = await self.search_symbols(query, limit, repository_path)

        if search_type in ("code", "both"):
            results["code"] = await self.search_code(query, limit, repository_path)

        return results

File: src/query.rs
//! Natural language query functionality

use crate::{database::Database, embeddings::EmbedderBackend, Config, Result};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

/// Symbol search result
///
/// Contains symbol metadata and similarity score from vector search.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SymbolResult {
    /// Symbol database ID
    pub id: i32,
    /// Symbol name
    pub name: String,
    /// Symbol kind ("function", "class", etc.)
    pub kind: String,
    /// Full signature (for functions/methods)
    pub signature: Option<String>,
    /// Documentation string
    pub docstring: Option<String>,
    /// Starting line number (1-indexed)
    pub start_line: i32,
    /// Ending line number (1-indexed)
    pub end_line: i32,
    /// File path
    pub path: String,
    /// Programming language
    pub language: Option<String>,
    /// Repository path
    pub repo_path: String,
    /// Repository name
    pub repo_name: String,
    /// Similarity score (0.0 to 1.0, higher is more similar)
    pub similarity: f64,
}

/// Natural language query engine
///
/// Performs semantic search over indexed code using embeddings.
pub struct QueryEngine {
    db: Arc<Database>,
    embedder: Arc<EmbedderBackend>,
}

impl QueryEngine {
    /// Create a new query engine
    ///
    /// # Arguments
    /// * `config` - Application configuration
    /// * `db` - Database connection
    ///
    /// # Returns
    /// Query engine ready to process searches
    pub fn new(config: Arc<Config>, db: Arc<Database>) -> Result<Self> {
        let embedder = Arc::new(EmbedderBackend::from_config(&config)?);

        Ok(QueryEngine { db, embedder })
    }

    /// Search for symbols using natural language query
    ///
    /// Converts the query to an embedding and performs vector similarity search.
    ///
    /// # Arguments
    /// * `query` - Natural language search query
    /// * `limit` - Maximum number of results
    /// * `repository_path` - Optional filter to specific repository
    ///
    /// # Returns
    /// Vector of symbol results ordered by similarity (most similar first)
    pub async fn search_symbols(
        &self,
        query: &str,
        limit: i64,
        repository_path: Option<&str>,
    ) -> Result<Vec<SymbolResult>> {
        let query_embedding = self.embedder.encode(query)?;

        // Debug: Check if embedding is valid
        if std::env::var("CUDGEL_DEBUG").is_ok() {
            eprintln!("Query: \"{}\"", query);
            eprintln!("Embedding length: {}", query_embedding.len());
            eprintln!(
                "First 5 values: {:?}",
                &query_embedding[..5.min(query_embedding.len())]
            );
            let norm: f32 = query_embedding.iter().map(|x| x * x).sum::<f32>().sqrt();
            eprintln!("Embedding L2 norm: {:.6}", norm);
        }

        let rows = self
            .db
            .search_symbols(&query_embedding, limit, repository_path)
            .await?;

        if std::env::var("CUDGEL_DEBUG").is_ok() {
            eprintln!("Query returned {} rows", rows.len());
            for (i, row) in rows.iter().enumerate().take(3) {
                let name: String = row.get("name");
                let similarity: f64 = row.get("similarity");
                eprintln!(
                    "  Result {}: {} (similarity: {:.6})",
                    i + 1,
                    name,
                    similarity
                );
            }
        }

        let results = rows
            .iter()
            .map(|row| SymbolResult {
                id: row.get("id"),
                name: row.get("name"),
                kind: row.get("kind"),
                signature: row.get("signature"),
                docstring: row.get("docstring"),
                start_line: row.get("start_line"),
                end_line: row.get("end_line"),
                path: row.get("path"),
                language: row.get("language"),
                repo_path: row.get("repo_path"),
                repo_name: row.get("repo_name"),
                similarity: row.get("similarity"),
            })
            .collect();

        Ok(results)
    }
}
